### PR TITLE
Create Page Object Input Providers

### PR DESCRIPTION
**Pull request by victor-torres**

### CODE REVIEW

File: scrapy_autoextract/providers.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from scrapy import Request
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy_poet.page_input_providers import PageObjectInputProvider
+
+
+class _Provider(PageObjectInputProvider):
+
+    page_type: ClassVar[str]
+    provided_class: ClassVar[Optional[Type]]
```
Comment (kmike) [inline scrapy_autoextract/providers.py:18]: would it be possible to use the current scrapy-poet providers API?

File: setup.py [hunk 2]
```diff
@@ -26,6 +26,13 @@ def get_version():
     long_description=open('README.rst').read(),
     url='https://github.com/scrapinghub/scrapy-autoextract',
     packages=find_packages(),
+    install_requires=[
+        # FIXME: uncomment after release
+        # 'autoextract-poet',
+        'scrapinghub-autoextract==0.3.0',
+        'scrapy',
```
Comment (kmike) [inline setup.py:33]: scrapy was not in install_requires on purpose; usually plugins don't declare their frameworks as dependencies. In past it was more important, because it could have triggered unintended upgrades of the framework (e.g. `pip install -U scrapy-autoextract` could have upgraded Scrapy if Scrapy is in install_requires). Now I think pip defaults not to do this, but still, I think it is safer not to have it here.

File: setup.py [hunk 3]
```diff
@@ -26,6 +26,13 @@ def get_version():
     long_description=open('README.rst').read(),
     url='https://github.com/scrapinghub/scrapy-autoextract',
     packages=find_packages(),
+    install_requires=[
+        # FIXME: uncomment after release
+        # 'autoextract-poet',
+        'scrapinghub-autoextract==0.3.0',
```
Comment (kmike) [inline setup.py:32]: usually pinning versions in install_requires is a bad practice, as it may lead to conflicts easily. E.g. we release  `scrapinghub-autoextract==0.3.1` which is compatible; someone installs scrapy-autoextract, and scrapinghub-autoextract gets downgraded for an user.

File: setup.py [hunk 4]
```diff
@@ -26,6 +26,13 @@ def get_version():
...
+        'scrapy',
+        'scrapy-poet==0.0.3',
```
Comment (kmike) [inline setup.py:34]:
```suggestion
        'scrapy-poet>=0.0.3',
```

File: setup.py (hunk 3 above)
Comment (kmike) [reply setup.py:32]: A rule of thumb: `install_requires` should exclude versions which are known not to work, not specify versions which are known to work.

File: scrapy_autoextract/providers.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        request = AutoExtractRequest(
+            url=self.request.url,
+            pageType=self.page_type,
+        )
+
+        try:
+            data = await request_raw(
+                [request.as_dict()],
+                api_key=self.settings.get('AUTOEXTRACT_USER'),
```
Comment (kmike) [inline scrapy_autoextract/providers.py:41]: I think it should take the same setting as the current scrapy-autoextract.

**Comment by kmike** [comment]
@victor-torres could you please add some docs?

File: scrapy_autoextract/providers.py (hunk 5 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:41]: This is where I took this key name from: https://github.com/scrapinghub/scrapy-autoextract/blob/master/scrapy_autoextract/middlewares.py#L50

Comment (kmike) [reply scrapy_autoextract/providers.py:41]: Ahh, that's perfect, sorry!

File: setup.py (hunk 3 above)
Comment (victor-torres) [reply setup.py]: Noted. `'scrapinghub-autoextract==0.3.0'` was replaced by `'scrapinghub-autoextract>=0.3.0'`.

File: setup.py (hunk 2 above)
Comment (victor-torres) [reply setup.py]: It's complicated. I'm removing the library from the project's dependencies but I personally believe it should be listed as a requirement because our library makes use of its code. Scrapy and our library are interdependent.

File: scrapy_autoextract/providers.py (hunk 1 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:21]: Done. Keep in mind we need to import scrapy-autoextract's providers module in order to effectively import its providers.

**Comment by victor-torres** [comment]
@kmike, thank you for your previous review. I think this pull request is ready for another shot.

File: scrapy_autoextract/providers.py (hunk 1 above)
Comment (kmike) [reply scrapy_autoextract/providers.py:21]: There is no provided_class in the current scrapy-poet - how does it work?

Regarding having to import it, what do you think about not using `@provides` decorator, but creating a helper which registers all available providers?

```py
from scrapy_poet.page_input_providers import register

def install():
    register(ArticleResponseDataProvider, AutoExtractArticleData)
    register(ProductResponseDataProvider, AutoExtractProductData)
```

And then document that one should put the following code to settings.py to be able to use scrapy-autoextract with page objects?

```py
import scrapy_autoextract.poet
scrapy_autoextract.poet.install()
```

File: scrapy_autoextract/providers.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        except Exception:
+            self.stats.inc_value(f"autoextract/{self.page_type}/error")
+            raise
+
+        self.stats.inc_value(f"autoextract/{self.page_type}/success")
+        return self.provided_class(data=data)
+
+
+@provides(AutoExtractArticleData)
+class ArticleResponseDataProvider(_Provider):
```
Comment (kmike) [inline scrapy_autoextract/providers.py:56]: It is no longer called ResponseData, so I think providers should be renamed as well, e.g.

```suggestion
class ArticleDataProvider(_Provider):
```

File: scrapy_autoextract/providers.py (hunk 1 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:21]: I had the same feeling about this decorator. I guess it will leave on the next versions. Let me try to implement something based on your suggestion.

File: scrapy_autoextract/providers.py (hunk 6 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: Done.

File: scrapy_autoextract/providers.py (hunk 1 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:21]: Done.

File: scrapy_autoextract/providers.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        except Exception:
+            self.stats.inc_value(f"autoextract/{self.page_type}/error")
+            raise
+
+        self.stats.inc_value(f"autoextract/{self.page_type}/success")
+        return self.provided_class(data=data)
+
+    @classmethod
+    def register(cls):
+        register(cls, cls.provided_class)
```
Comment (victor-torres) [inline scrapy_autoextract/providers.py:56]: @kmike, I've created this auxiliary class method to avoid duplicating provided class name on the `install` procedure. 

Another alternative would be using the class attribute directly in the `install` procedure. Let me know if you have a preference for this other option and I'll refactor it removing this  `register` class method.

File: scrapy_autoextract/providers.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class ArticleDataProvider(_Provider):
+
+    page_type = "article"
```
Comment (victor-torres) [inline scrapy_autoextract/providers.py:61]: @kmike, we could read the page type straight from the provided class here, but we've just made it private. Example:

```python
AutoExtractArticleData._item_key
```

This would avoid duplication whilst making it more difficult to create custom providers maybe.

**Comment by victor-torres** [comment]
@kmike, I think it's a good time for a last check here.

File: scrapy_autoextract/providers.py (hunk 8 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: Update: this class attribute was made public again and we're using it to acquire our page type.

File: scrapy_autoextract/providers.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        self.stats.inc_value(f"autoextract/{page_type}/total")
+
+        request = AutoExtractRequest(
+            url=self.request.url,
+            pageType=page_type,
+        )
+
+        try:
+            data = await request_raw(
+                [request.as_dict()],
```
Comment (kmike) [inline scrapy_autoextract/providers.py:52]: as_dict looks unnecessary here, scrapinghub-autoextract should support Request objects

File: scrapy_autoextract/providers.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                api_key=self.settings.get('AUTOEXTRACT_USER'),
+                max_query_error_retries=3
```
Comment (kmike) [inline scrapy_autoextract/providers.py:54]: would you mind making this a Scrapy setting?

**Comment by kmike** [comment]
I left a couple of minor comments, but the code looks good @victor-torres. Could you please update the README, describing how to use this inteegration approach, from the user point of view?

File: scrapy_autoextract/providers.py (hunk 9 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: That's true! Done.

File: scrapy_autoextract/providers.py (hunk 10 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: Done. It's called `AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES` and it's described on the readme file.

**Comment by victor-torres** [comment]
Since the last review:

- I've removed the unnecessary call to `as_dict()` on the `Request` object
- I've introduced a new setting called `AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES`
- I've updated the base provider to make use of the `AUTOEXTRACT_URL` settings 
  - Because of that, we need to release a new version of [scrapinghub-autoextract](https://github.com/scrapinghub/scrapinghub-autoextract), 0.5.0 probably

Could you take another look, @kmike? Especially at the new readme file.

File: README.rst [hunk 11]
```diff
@@ -42,19 +63,46 @@ Add the AutoExtract downloader middleware in the settings file::
...
+
+    from autoextract_poet.page_inputs import AutoExtractArticleData
+
+    def parse(self, response, article: AutoExtractArticleData):
+        yield article.to_item()
+
+Configuration
+^^^^^^^^^^^^^
+
+Make sure you have `scrapy-poet`_ properly configured.
```
Comment (kmike) [inline README.rst:85]: Currently to unuderstand how to configure scrapy-poet user would need to 

* go to https://github.com/scrapinghub/scrapy-poet
* Scroll down to the README, click on "Documentation", get to https://scrapy-poet.readthedocs.io/en/latest/
* go to "Installation guide" (it had no information)
* go to Tutorial, and read it - only there, somewhere in the middle there was a section on how to configure scrapy-poet properly (https://scrapy-poet.readthedocs.io/en/latest/intro/tutorial.html#configuring-the-project)

What do you think about

1) changing scrapy-poet docs to make configuration much more upfront,
2) pointing to this section directly here?

It was also unclear to me if "Make sure you have `scrapy-poet`_ properly configured." is done by the example code below, if if I should try find something in scrapy-poet docs. Maybe we can change the wording a bit, e.g. "First, configure scrapy-poet, as described here. Then, enable AutoExtract providers, by putting the following code to Scrapy's settinigs.py file:".

File: README.rst [hunk 12]
```diff
@@ -67,11 +115,10 @@ Available settings:
...
-Within the spider, consuming the AutoExtract result is as easy as::
-
-    def parse(self, response):
-        yield response.meta['autoextract']
+Provider settings
+-----------------
 
+- ``AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES`` [optional] Max number of retries for Query-level errors. Defaults to ``3``.
 
 Limitations
```
Comment (kmike) [inline README.rst:123]: I think it'd be nice to mention some of the limitations of the Scrapy middleware, as compared to the scrapy-poet provider:
* 429 errors are handled as standard retries when using Scrapy middleware, but they're handled properly&automatically with scrapy-poet, as it relies on scrapinghub-autoextract. One may loose some responses with the middleware. With scrapy-poet there is no need to change RETRY_HTTP_CODES.
* retrying of query-level errors is only supported in scrapy-poet integration (?)
* retries overall have a better behavior with scrapy-poet integration
* with scrapy-poet integration retry requests don't go through Scrapy
* not all data types are supported with scrapy-poet, currently only article and product are supported

File: scrapy_autoextract/providers.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            pageType=page_type,
+        )
+        api_key = self.settings.get("AUTOEXTRACT_USER")
+        endpoint = self.settings.get("AUTOEXTRACT_URL")
+        max_query_error_retries = self.settings.getint(
+            "AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES", 3
+        )
+
+        try:
+            data = await request_raw(
```
Comment (kmike) [inline scrapy_autoextract/providers.py:56]: I wonder if you need to set twisted reactor to asyncio for this to work, or now. If there is non-default Scrapy configuration enabled, we should mention it in readme.

File: README.rst (hunk 11 above)
Comment (victor-torres) [reply README.rst]: This is a very nice observation. I've just implemented your suggestion:

- rephrased this part of our documentation
- included a link that takes the user straight to the configuration docs of scrapy-poet

Even then, I think there's room for improvement in scrapy-poet's documentation.
Since this is kinda out of the scope of this task, I've created a new issue there: https://github.com/scrapinghub/scrapy-poet/issues/36

File: README.rst (hunk 12 above)
Comment (victor-torres) [reply README.rst:122]: Those are good comments that are worth mentioning in our limitations section. I've just pushed a new commit adding this information. Could you check them, please? If you have further suggestions or would like to change the way something is written, please, feel free to submit a commit to this branch making these changes to avoid back and forth.

File: scrapy_autoextract/providers.py (hunk 13 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:56]: Before merging this pull request, let me perform a last test with a sample spider just to make sure everything is working as expected, okay? I should get back to you tomorrow afternoon.

File: scrapy_autoextract/providers.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+                [request],
+                api_key=api_key,
+                endpoint=endpoint,
+                max_query_error_retries=max_query_error_retries
+            )
+        except Exception:
+            self.stats.inc_value(f"autoextract/{page_type}/error")
+            raise
+
+        self.stats.inc_value(f"autoextract/{page_type}/success")
```
Comment (ivanprado) [inline scrapy_autoextract/providers.py:66]: We should check here for the content. If `error` in `data[0]` then raise `QueryLevelError`.  Only True successful requests should pass without exception.

File: scrapy_autoextract/providers.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+
+class _Provider(PageObjectInputProvider):
+    """An interface that describes a generic AutoExtract Provider.
+
+    It should not be used publicly as it serves the purpose of being a base
+    class for more specific providers such as Article and Product providers.
+    """
+
+    provided_class: ClassVar[Optional[Type]]
```
Comment (ivanprado) [inline scrapy_autoextract/providers.py:25]: Why `Optional`? Shouldn't be always there?

Comment (ivanprado) [reply scrapy_autoextract/providers.py:25]: The type must be a subclass of `_AutoExtractData` ensuring that item_key is declared. It could be nice to have a comment here stating that.

File: scrapy_autoextract/providers.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        based on API response data.
+        """
+        page_type = self.get_page_type()
+        self.stats.inc_value(f"autoextract/{page_type}/total")
+
+        request = AutoExtractRequest(
+            url=self.request.url,
+            pageType=page_type,
+        )
+        api_key = self.settings.get("AUTOEXTRACT_USER")
```
Comment (ivanprado) [inline scrapy_autoextract/providers.py:49]: Here we force to define that in the settings, but it would be good that by default it can be picked up from `SCRAPINGHUB_AUTOEXTRACT_KEY` environment variable, as is done in https://github.com/scrapinghub/scrapinghub-autoextract. 

That way we would have a way of configuring the key a single time and use it from all the Scrapinghub tools.

In other words, if `AUTOEXTRACT_USER` is not in the settings, then just rely on the ENV variable. 

Also probably I would change the name to `AUTOEXTRACT_KEY` to be coherent.

**Review by ivanprado — COMMENTED** [review]
@victor-torres good stuff in this PR :+1: I've left some comments.

File: scrapy_autoextract/providers.py (hunk 16 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:49]: - The reason we're naming it `AUTOEXTRACT_USER` instead of `AUTOEXTRACT_KEY` is to keep it consistent with the middleware (see [here](https://github.com/scrapinghub/scrapy-autoextract/blob/master/scrapy_autoextract/middlewares.py#L50)).
- The `api_key` argument is an optional string that fallbacks to the environment variable  `SCRAPINGHUB_AUTOEXTRACT_KEY` if it's `None` (see [here](https://github.com/scrapinghub/scrapinghub-autoextract/blob/9e7f3cc03208d54b858708b8f556acf7d312335b/autoextract/apikey.py#L12)).

Comment (victor-torres) [reply scrapy_autoextract/providers.py:49]: You could propose the change from `AUTOEXTRACT_USER` to `AUTOEXTRACT_KEY` in a new issue, what do you think, @ivanprado?

File: scrapy_autoextract/providers.py (hunk 15 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: > Why Optional? Shouldn't be always there?

It shouldn't be optional. I've just fixed it. Thanks.

File: scrapy_autoextract/providers.py (hunk 14 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:77]: Nice observation, Iván! I've just refactored this code to raise an exception when the response data contains an error.

File: scrapy_autoextract/providers.py (hunk 15 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: > The type must be a subclass of _AutoExtractData ensuring that item_key is declared.

We could do something like `provided_class: ClassVar[Type[_AutoExtractData]]` if this class was not private. 

> It could be nice to have a comment here stating that.

I guess I'm going with this option. Let me know if you, @kmike or @ejulio have another option.

Comment (victor-torres) [reply scrapy_autoextract/providers.py]: What do you think? Should we make `_AutoExtractData` public and use it here for describing `provided_class`'s type?

Comment (kmike) [reply scrapy_autoextract/providers.py]: +1 to add a comment

File: scrapy_autoextract/providers.py (hunk 16 above)
Comment (kmike) [reply scrapy_autoextract/providers.py:54]: +1 to discuss rename as a separate issue. Regarding ENV variable, if this works, it'd be good to document that, and check (at least manually).

File: scrapy_autoextract/providers.py (hunk 15 above)
Comment (ivanprado) [reply scrapy_autoextract/providers.py]: +1 to add a comment.

File: scrapy_autoextract/providers.py (hunk 16 above)
Comment (ivanprado) [reply scrapy_autoextract/providers.py:54]: Agree.

File: scrapy_autoextract/providers.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+from scrapy import Request
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy_poet.page_input_providers import (
+    PageObjectInputProvider,
+    register,
+)
+
+
+class QueryLevelError(Exception):
```
Comment (ivanprado) [inline scrapy_autoextract/providers.py:18]: It would be nice to get the error message isolated. See for example this implementation I have in other code:

```py
class QueryLevelError(Exception):
    def __init__(
        self, url: str, page_type: str, query: List, response: Dict, error_msg: str
    ):
        self.url = url
        self.page_type = page_type
        self.query = query
        self.error_msg = error_msg

    def __str__(self):
        return (
            f"QueryLevelError: url={self.url}, page_type={self.page_type}, "
            f"error={self.error_msg}, query:{self.query}"
        )
```

Probably not all this information is required, but I would keep at least the error and the query.

File: scrapy_autoextract/providers.py (hunk 16 above)
Comment (ivanprado) [reply scrapy_autoextract/providers.py:54]: Created an issue here: https://github.com/scrapinghub/scrapy-autoextract/issues/14

Comment (victor-torres) [reply scrapy_autoextract/providers.py:54]: > Regarding ENV variable, if this works, it'd be good to document that, and check (at least manually).

It works because my sample spider does not define `AUTOEXTRACT_USER` on `settings.py`.
I've updated the readme file describing this setting as optional when using providers and mentioning `SCRAPINGHUB_AUTOEXTRACT_KEY` environment variable as its default value.

File: scrapy_autoextract/providers.py (hunk 17 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py]: Done, `QueryError` exception was improved.

**Review by ivanprado — COMMENTED** [review]
@victor-torres It looks great! :+1: 

Only a final thing from my side. We should have some documentation about which exceptions are raised and why. In this case, there are two possible exceptions `RequestError` and `QueryError`. Not sure if this is the place to document them, or maybe is instead `scrapy-poet`.

File: scrapy_autoextract/providers.py [hunk 18]
```diff
@@ -72,9 +77,7 @@ async def __call__(self):
 
         if "error" in data:
             self.stats.inc_value(f"autoextract/{page_type}/error/query")
-            raise QueryLevelError(
-                f"Error '{data['error']}' while processing {self.request}"
-            )
+            raise QueryError(data["query"], data["error"])
```
Comment (kmike) [inline scrapy_autoextract/providers.py:80]: What's you opinion on using scrapinghub-autoextract QueryError object instead (see https://github.com/scrapinghub/scrapinghub-autoextract/blob/9e7f3cc03208d54b858708b8f556acf7d312335b/autoextract/aio/errors.py#L60)?

Is it possible for two different types of QueryErrors to propagate to the user code? If it is, having them by the same name can be confusing.

File: README.rst [hunk 19]
```diff
@@ -94,3 +167,23 @@ When using the AutoExtract middleware, there are some limitations.
   There is an exception, if there are too many requests sent in
   a short amount of time and AutoExtract returns HTTP code 429.
   For that case it's best to use ``RETRY_HTTP_CODES=[429]``.
+* 429 errors are handled as standard retries when using Scrapy middleware,
+  but they're handled properly and automatically with scrapy-poet integration,
+  as it relies on `scrapinghub-autoextract`_.
+  You may loose some responses with the middleware.
```
Comment (kmike) [inline README.rst:173]:
```suggestion
  You may lose some responses with the middleware.
```

File: scrapy_autoextract/providers.py (hunk 18 above)
Comment (victor-torres) [reply scrapy_autoextract/providers.py:80]: > What's you opinion on using scrapinghub-autoextract QueryError object instead

I don't think mixing up exceptions from two different libraries is a good idea. 

> Is it possible for two different types of QueryErrors to propagate to the user code?

No, QueryError exceptions on scrapinghub-autoextract are captured by the request method. When it happens, the latest results available are returned to the caller. A good idea is to turn this exception private to avoid confusion since it's not supposed to end-users.

Comment (victor-torres) [reply scrapy_autoextract/providers.py:80]: https://github.com/scrapinghub/scrapinghub-autoextract/issues/18

File: README.rst [hunk 20]
```diff
@@ -42,19 +63,45 @@ Add the AutoExtract downloader middleware in the settings file::
...
-or just for specific spiders, in the ``custom_settings`` dict.
+Page Objects their returned Items are defined by the `autoextract-poet`_
+library.
+
+Within the spider, consuming the AutoExtract result is as easy as::
+
+    from autoextract_poet.page_inputs import AutoExtractArticleData
+
+    def parse(self, response, article: AutoExtractArticleData):
+        yield article.to_item()
```
[the remark this answers is not in the document]
Comment (ejulio) [reply README.rst:80]: I'd add the spider code around it as well.
It would be 5 other lines, but would make it easier to link to a request and the parse and also making it explicit that a single request is being downloaded twice

File: README.rst [hunk 21]
```diff
@@ -42,16 +63,67 @@ Add the AutoExtract downloader middleware in the settings file::
...
+
+Note that on the example above, we're going to perform two requests:
+
+* one goes through Scrapy (it might use Crawlera, Splash or no proxy at all, depending on your configuration)
+* another goes through AutoExtract API using `scrapinghub-autoextract`_
+
+If you don't need the additional request going through Scrapy,
+you can annotate the response argument of your callback with the DummyResponse type.
```
Comment (ejulio) [inline README.rst:88]: ``DummyResponse``

File: setup.py [hunk 22]
```diff
@@ -26,6 +26,11 @@ def get_version():
     long_description=open('README.rst').read(),
     url='https://github.com/scrapinghub/scrapy-autoextract',
     packages=find_packages(),
+    install_requires=[
+        'autoextract-poet>=0.0.1',
```
Comment (ejulio) [inline setup.py:30]: I wonder if these should be installed by default, even if the user does not want it.
It can help to reduce the end package size, on the user side.
Also, we might want to use `scrapy-autoextract` in a python version that doesn't support asyncio, so not sure if it can be a problem

File: README.rst [hunk 23]
```diff
@@ -89,8 +162,24 @@ When using the AutoExtract middleware, there are some limitations.
...
-  For that case it's best to use ``RETRY_HTTP_CODES=[429]``.
+* AutoExtract doesn't handle retries internally, they should be handled by its clients.
+* 429 errors could be handled as standard retries when using Scrapy middleware,
+  but they're handled properly and automatically with scrapy-poet integration,
+  as it relies on `scrapinghub-autoextract`_.
+  You may lose some responses with the middleware approach.
+* Overall, retries have a better behavior with scrapy-poet integration
+  and it includes support for automatic Query-level errors retries with
+  no need to change ``RETRY_HTTP_CODES``.
+
```
Comment (victor-torres) [inline README.rst:173]: @kmike, would you mind taking another look at this section? Feel free to suggest more improvements.

File: README.rst (hunk 21 above)
Comment (victor-torres) [reply README.rst:88]: Sorry, what's supposed to be changed here?

File: README.rst [hunk 24]
```diff
@@ -42,16 +63,67 @@ Add the AutoExtract downloader middleware in the settings file::
...
+If you don't need the additional request going through Scrapy,
+you can annotate the response argument of your callback with the DummyResponse type.
+This will ignore the Scrapy request and only the AutoExtract API will be fetched.
+
+For example::
+
+    from autoextract_poet.page_inputs import AutoExtractArticleData
+    from scrapy_poet.utils import DummyResponse
+
+    def parse(self, response: DummyResponse, article: AutoExtractArticleData):
```
[the remark this answers is not in the document]
Comment (victor-torres) [reply README.rst:96]: > I wonder if we can have it as [...] just to have the more important arguments

We can, but I think we should keep it consistent with the regular callback signature that specifies the response as the first argument.

> and maybe response can be renamed to _ if it is not used

I like to make it explicit what we're receiving through arguments even if the variable ends up not being used.

File: setup.py (hunk 22 above)
Comment (victor-torres) [reply setup.py:30]: > I wonder if these should be installed by default, even if the user does not want it.

You mean, if users are going to use the middleware approach, this library will be installed as if they were going to use providers, right?

Well, I'm suspect and probably biased, but I believe libraries should include all its dependencies even if users end up not using all of them. It represents less trouble and makes it easier for users when they decide it's time to try the other way around. 

The alternative would be something like we have/used to have on Spidermon, something like `pip install spidermon[monitoring]` but I'm not a big fan of this approach.

> It can help to reduce the end package size, on the user side.

I think the difference in package size would be negligible.

> Also, we might want to use scrapy-autoextract in a python version that doesn't support asyncio, so not sure if it can be a problem

I'm not sure it would be a good idea to use those Scrapy providers in a sync way because it could block our requests cycle.

File: README.rst (hunk 21 above)
Comment (ejulio) [reply README.rst]: DummyResponse is not formatted properly 
Forgot that it would be formatted in the comment :sweat_smile:

File: README.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (ejulio) [reply README.rst]: > I like to make it explicit what we're receiving through arguments even if the variable ends up not being used.

You are being explicit, by typing it as `DummyResponse`.

But I agree with your thoughts, keeping the consistency might be better in this case

File: setup.py (hunk 22 above)
Comment (ejulio) [reply setup.py:30]: > I'm not sure it would be a good idea to use those Scrapy providers in a sync way because it could block our requests cycle. In any case, I believe asyncio is supported at least since Python 3.6, and Python 3.5 is currently deprecated.

Yeah, but what if I just want to use the middleware version?
And I'm in version 3.4 and a new version of this library is released with a fix in the middleware?

> I think the difference in package size would be negligible.

No hard feelings on this one.
Just that we count the popularity of a package by downloads :stuck_out_tongue: 

About the installation, it will depend on the first case, if we want to install async things by default, even if it may break future fixes..

File: README.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (ivanprado) [reply README.rst]: Scrapy forces the first argument to be the response, and this is something that the the `InjectMiddelware` cannot change. So it is not possible to change the order of the parameters in this example.

File: README.rst [hunk 25]
```diff
@@ -42,16 +63,137 @@ Add the AutoExtract downloader middleware in the settings file::
...
+^^^^^^^^^^
+
+While trying to fetch AutoExtract API, providers might raise some exceptions.
+Those exceptions might come from scrapy-autoextract providers themselves,
+`scrapinghub-autoextract`_, or Tenacity, the library used to implement retries.
+For example:
+
+* ``scrapy_poet.providers.QueryError``: raised when a `Query-level error`_ is returned even after retrying the request
+* ``autoextract.aio.errors.RequestError``: raised when a `Request-level error`_ is returned
+* ``autoextract.aio.errors.QueryRetryError``: raised when it's not possible to retry a `Query-level error`_
```
Comment (ivanprado) [inline README.rst:155]: I think this one is never raised. Should be removed from here.

File: README.rst [hunk 26]
```diff
@@ -42,16 +63,137 @@ Add the AutoExtract downloader middleware in the settings file::
...
+* ``tenacity.RetryError``: raised when it's not possible to retry an error
```
Comment (ivanprado) [inline README.rst:156]: IMO we should change the retrial code adding the parameter `reraise=True` to avoid raising these exceptions at all. See https://tenacity.readthedocs.io/en/latest/#error-handling. 

Probably better to do so in a different PR

File: README.rst (hunk 25 above)
Comment (victor-torres) [reply README.rst:154]: I've removed the line that cites QueryError exceptions because they never reach userland's code. 
QueryRetryError does.

More details on the related commit's message.

File: README.rst (hunk 26 above)
Comment (victor-torres) [reply README.rst]: Would you mind creating a new issue for that?

Comment (ivanprado) [reply README.rst]: Sure. Created here: https://scrapinghub.atlassian.net/browse/TRIN-103

File: README.rst (hunk 25 above)
Comment (ivanprado) [reply README.rst:154]: @victor-torres I think that `QueryRetryError` also never reaches the user. See this lines of code: https://github.com/scrapinghub/scrapinghub-autoextract/blob/master/autoextract/aio/client.py#L259-L263

The exception is captured, and just the last result is returned without reraising the exception. 

But maybe I'm missing something.

Comment (ivanprado) [reply README.rst:154]: This seems consistent with the `request_raw` documentation: https://github.com/scrapinghub/scrapinghub-autoextract/blob/master/autoextract/aio/client.py#L157-L163

File: README.rst (hunk 26 above)
Comment (ivanprado) [reply README.rst]: It seems the issue was already there: https://github.com/scrapinghub/scrapinghub-autoextract/issues/11

I'll update it.

File: README.rst (hunk 25 above)
Comment (victor-torres) [reply README.rst]: That's very true. We've discussed this during our call the other day but I totally forgot about it. Thanks for pointing it. Fixed/removed.

**Review by ivanprado — COMMENTED** [review]
I gave another look

File: README.rst [hunk 27]
```diff
@@ -42,16 +63,135 @@ Add the AutoExtract downloader middleware in the settings file::
...
+                yield scrapy.Request(url, callback=self.parse_article, errback=self.errback_article)
+
+        def parse_article(self, response: DummyResponse, article: AutoExtractArticleData):
+            yield article.to_item()
+
+        def errback_article(self, failure: Failure):
+            if failure.check(RequestError):
+                self.logger.error(f"RequestError on {failure.request.url})
+
+            if failure.check(QueryRetryError):
```
[the remark this answers is not in the document]
Comment (victor-torres) [reply README.rst:181]: For sure. Removed.

**Review by ivanprado — APPROVED** [review]
Thank you @victor-torres! Great code :+1: It looks good to me.

File: README.rst [hunk 28]
```diff
@@ -89,8 +227,27 @@ When using the AutoExtract middleware, there are some limitations.
   so it's best to use ``AUTHTHROTTLE_ENABLED=False`` in the settings.
 * Redirects are handled by AutoExtract, not by Scrapy,
   so these kinds of middlewares might have no effect
-* Retries should be disabled, because AutoExtract handles them internally
-  (use ``RETRY_ENABLED=False`` in the settings)
-  There is an exception, if there are too many requests sent in
-  a short amount of time and AutoExtract returns HTTP code 429.
-  For that case it's best to use ``RETRY_HTTP_CODES=[429]``.
+* AutoExtract doesn't handle retries internally, they should be handled by its clients.
```
Comment (kmike) [inline README.rst:230]: Let's remove this line. It is an implementation detail which can change, and it is not a limitation of scrapy-autoextract, it is an API behavior.

**Comment by kmike** [comment]
@victor-torres looks good to me, thanks! I only left a minor comment about docs.

**Review by kmike — APPROVED** [review]

File: README.rst (hunk 28 above)
Comment (victor-torres) [reply README.rst]: Removed.

**Comment by victor-torres** [comment]
Thank you for your attention and kind reviews. I'm merging this pull request now.
